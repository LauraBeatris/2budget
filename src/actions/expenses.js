import uuid from "uuid";

//ADD_EXPENSE
export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0
} = {}) => {
  return {
    type: "ADD_EXPENSE",
    expense: {
      // Setting a id
      id: uuid(),
      description,
      note,
      amount,
      createdAt
    }
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

//EDIT_EXPENSE
export const editExpense = (id, update) => ({
  type: "EDIT_EXPENSE",
  id,
  update
});
