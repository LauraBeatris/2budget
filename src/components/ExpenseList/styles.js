import styled from "styled-components";

export const ListContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
  font-family: "Montserrat", sans-serif;
  margin-top: 4rem;

  .summary {
    line-height: 1;
    margin-bottom: 4rem;
    color: #6dbfcc;
    h1.total-amount {
      font-weight: 400;
      font-size: 6.5rem;
      margin-bottom: 1.5rem;
    }

    h2 {
      font-size: 2rem;
      text-transform: uppercase;
      font-weight: bold;
    }

    .moment, .quantity {
      font-size: 1.8rem;

      margin-bottom: 1rem;
      text-transform: uppercase;
      font-weight: 500;
    }
  }

  @media screen and (max-width: 45rem) {
    margin: 2rem auto;
  }
`;
