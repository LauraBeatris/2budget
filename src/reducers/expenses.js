const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.update // overriden the expense items with the action updates icons
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
