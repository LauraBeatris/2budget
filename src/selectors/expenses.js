import moment from "moment";

// Get visible expenses
export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return b.createdAt < a.createdAt ? -1 : 1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? -1 : 1;
      } else {
        return null;
      }
    });
};

export const getExpenseTotal = expenses => {
  if (expenses.length === 0) {
    return 0;
  } else if (expenses.length > 1) {
    const amounts = expenses.map(expense => expense.amount);
    return amounts.reduce((sum, value) => sum + value, 1);
  } else {
    return expenses[0].amount;
  }
};
