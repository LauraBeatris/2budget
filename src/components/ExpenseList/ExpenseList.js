import React from "react";
import { connect } from "react-redux";
import ExpenseItem from "../ExpenseItem/ExpenseListItem";
import selectExpenses from "../../selectors/expenses";
import { ListContainer } from "./styles";
import moment from "moment";

// Regular unconnect components
export const ExpenseList = props => (
  <ListContainer>
    <div className="summary">
      <p className="moment">{moment().format("MMMM YYYY")}</p>
      <h1 className="total-amount">$208,5</h1>
      <h2>Total Amount</h2>
    </div>
    {props.expenses.length === 0 ? (
      <p>No expenses</p>
    ) : (
      props.expenses.map(expense => (
        <ExpenseItem key={expense.id} {...expense} />
      ))
    )}
  </ListContainer>
);

// Mapping the state to the props of the component
const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

// Exporting the HOC with the props
export default connect(mapStateToProps)(ExpenseList);
