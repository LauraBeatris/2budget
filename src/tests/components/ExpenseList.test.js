import React from "react";
import { ExpenseList } from "../../components/ExpenseList/ExpenseList";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";

test("should render ExpenseList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseList with empty message", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
