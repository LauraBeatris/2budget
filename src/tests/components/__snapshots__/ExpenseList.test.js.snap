// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<.sc-bwzfXH>
  <div
    className="summary"
  >
    <p
      className="moment"
    >
      December 1969
    </p>
    <h1
      className="total-amount"
    >
      $208,5
    </h1>
    <h2>
      Total Amount
    </h2>
  </div>
  <p>
    No expenses
  </p>
</.sc-bwzfXH>
`;

exports[`should render ExpenseList with expenses 1`] = `
<.sc-bwzfXH>
  <div
    className="summary"
  >
    <p
      className="moment"
    >
      December 1969
    </p>
    <h1
      className="total-amount"
    >
      $208,5
    </h1>
    <h2>
      Total Amount
    </h2>
  </div>
  <ExpenseItem
    amount={195}
    createdAt={"1970-01-01T00:00:00.000Z"}
    description="Water rent"
    id="1"
    key="1"
    note="Random note"
  />
  <ExpenseItem
    amount={3100}
    createdAt={-345600000}
    description="Gas rent"
    id="2"
    key="2"
    note="Random note"
  />
  <ExpenseItem
    amount={31000}
    createdAt={345600000}
    description="College rent"
    id="3"
    key="3"
    note="Random note"
  />
</.sc-bwzfXH>
`;
